### ISSUE TITLE
Support Debugging in Mono scripting

### ISSUE THREAD

**Issue by Redth** [cake-build/cake#1559]
I wanted to start a recorded discussion here of what it might take to support debugging when running under Mono via mono's remote debugger.   The main use case I am interested in is being able to debug through VS-Code's Mono Debugger extension.

So far I think I have the pieces to make this happen from VS-Code, but we're obviously still missing support in Cake's implementation that uses the mono script host.

Here's the piece for VS-Code that I think should mostly make it work once Cake supports it:
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "StartCake",
            "type": "mono",
            "request": "launch",
            "program": "${workspaceRoot}/tools/Cake/Cake.exe",
            "runtimeArgs": [
                "--debug",
                "--debugger-agent=transport=dt_socket,server=y,address=127.0.0.1:55556"
            ],
            "args": [
                "${workspaceRoot}/build.cake",
                "--verbosity=diagnostic",
                "--debug"
            ],
            "cwd": "${workspaceRoot}",
            "console":"internalConsole",
            "noDebug": true
        },
       {
            "name": "AttachCake",
            "type": "mono",
            "request": "attach",
            "address": "localhost",
            "port": 55556
        }
    ],
     "compounds": [
        {
            "name": "DebugCake",
            "configurations": ["StartCake", "AttachCake"]
        }
    ]
}
```

As for the Cake changes, I think first of all we need to take a look somewhere in here:
https://github.com/cake-build/cake/blob/develop/src/Cake/Scripting/Mono/MonoScriptSession.cs#L54

Basically, this SO answer shows how to enable generating debug symbols in mono's script host:
http://stackoverflow.com/questions/875723/how-to-debug-break-in-codedom-compiled-code

What I'm not exactly clear on is how do we go from having debug symbols generated for the generated script, to resolving breakpoints from the original script before it was transformed...

@mholo65  any ideas? I see you implemented the original debugging on the Roslyn side...

**Comment by mholo65** [comment]
On the Roslyn side (https://github.com/cake-build/cake/blob/develop/src/Cake/Scripting/Roslyn/Stable/DebugRoslynScriptSession.cs#L41) we emit the compilation to memorystreams (one for dll, and one for the pdb). Then we load the dll (+pdb) into the appdomain, and invoke "script" via reflection.

Resolving breakpoints works because scriptgen adds #Line preprocessor directives to the generated script. The #Line directives links back to the original Cake file.

**Comment by mholo65** [comment]
@patriksvensson may correct me if I'm wrong, but in v0.20.0 when Roslyn is updated (https://github.com/cake-build/cake/issues/1540) there will no longer be need to use separate mono scripting in Cake? We can finally use an unified scripting API?

**Comment by devlead** [comment]
@mholo65 0.20.0 will be updating Cake to VS2017 / new RTM .NET Core SDK, but correct for 0.21.0 that's the plan.

**Comment by mholo65** [comment]
@devlead thanks! I see milestone have been updated..

@redth, this practically means that there's no use on spending time on Mono specific scripting bits in Cake, as they will be obsolete/removed in v0.21.0

@devlead or @patriksvensson can correct me if I'm wrong :smile:

**Comment by devlead** [comment]
@mholo65 hopefully upgrading Roslyn is smooth sailing and works well with mono, then all the mono scriptgen stuff can be yanked out.

Don't know exact status of 0.20.0 but believe @patriksvensson said it was mainly yak shaving like StyleCop, build script refactoring, etc.

**Comment by patriksvensson** [comment]
@devlead @mholo65 Correct.

**Comment by Redth** [comment]
Niiiice. Hopefully then debugging with what I've attempted in vs code will just work at that point!

**Comment by mholo65** [comment]
Yes, in v0.21.0 it ***should*** work out of the box.

**Comment by patriksvensson** [comment]
@Redth Can we close this issue?

**Comment by Redth** [comment]
Yeah. I'll open one in the vs-code repo to track getting mono debugging working there.
